iiasa/message-ix-models
# Update sdmx dependency & migrate pyproject.toml to poetry for inspection

Closes #129.

As pointed out by @macflo8, the sdmx dependency version needs to be bumped. And all of the below can be discarded by dropping the second and third commit, if we don't like it. However:
While working on this, I found myself thinking that I'd like to manage dependencies with [poetry](https://python-poetry.org/docs/) as we already do in ixmp4. With poetry, we could quite easily update more dependencies as well: 
```console
$ poetry show -o
```
shows all packages in the venv that can be updated and whether this update is considered safe; 
```console
$ poetry show -o -T
```
shows all of these packages that we explicitly specify in our `pyproject.toml`. In this case, the output of the latter is
```shell
pyam-iamc                     1.9.0    2.0.0     Analysis & visualization of integrated-assessment sc...
sphinx                        6.2.1    7.2.6     Python documentation generator
sphinxcontrib-bibtex          2.5.0    2.6.1     Sphinx extension for BibTeX style citations.
```
With pip, I would now specify `pyam-iamc >= 2.0.0` in the dependencies to force the new version. In contrast, poetry can manage `pyproject.toml` for us and keeps us from making that mistake:
```console
$ poetry add pyam-iamc@latest
Using version ^2.0.0 for pyam-iamc

Updating dependencies
Resolving dependencies... (0.0s)

The current project's Python requirement (>=3.8,<3.12) is not compatible with some of the required packages Python requirement:
  - pyam-iamc requires Python <3.12,>=3.10, so it will not be satisfied for Python >=3.8,<3.10

Because no versions of pyam-iamc match >2.0.0,<3.0.0
 and pyam-iamc (2.0.0) requires Python <3.12,>=3.10, pyam-iamc is forbidden.
So, because message-ix-models depends on pyam-iamc (^2.0.0), version solving failed.

  • Check your dependencies Python requirement: The Python requirement can be specified via the `python` or `markers` properties
    
    For pyam-iamc, a possible solution would be to set the `python` property to ">=3.10,<3.12"
```
So we immediately know that we will only be able to use pyam-iamc 2.0.0 once we move our python requirement to be at least 3.10.

Poetry has some other nice features such as a focus on environment isolation. If poetry detects it is running in a venv, it will use that one, but if it isn't running in one, it will create and manage one for you. For ixmp4, we use the config option `poetry config virtualenvs.in-project true` to have these venvs installed in the project rather than globally. 

On the user side, nothing should change by including this version of `pyproject.toml` and `poetry.lock`. We can still run
```console
$ pip install message-ix-models[docs,report,tests]
```
or equivalently, when using poetry,
```console
# To not remember all extra-names
$ poetry install --all-extras

# To specify only some extras
$ poetry install --extras "docs tests"
```

However, it might be possible to improve other things as well: maybe we can use the automatically created `poetry.lock` as an input file for pinned dependencies for the docs builds that is automatically kept up-to-date when we update dependencies via poetry.

To install message-ix-models using poetry, you can follow these steps from within your local clone of this branch:
```console
# Install Poetry, minimum version >=1.2 required
$ curl -sSL https://install.python-poetry.org | python -

# You may have to reinitialize your shell at this point.
$ source ~/.bashrc

# Activate in-project virtualenvs
$ poetry config virtualenvs.in-project true

# Add dynamic versioning plugin 
$ poetry self add "poetry-dynamic-versioning[plugin]"

# Install dependencies
$ poetry install --all-extras
```

Once this is done, you can access the venv from the directory by running `poetry shell`, which will spawn a new shell where you can run `mix-models` and everything else as usual. However, you can also use `poetry run <command>` to [run cli commands](https://python-poetry.org/docs/basic-usage/#using-poetry-run) without creating a new shell.

Please note that I dropped some classifiers since poetry is adding [python and license classifiers automatically](https://python-poetry.org/docs/pyproject/#classifiers).

## How to review

- Read the diff and note that the CI checks all pass.
- Try installing the project via poetry and see how you like it :)
- Ensure that changes/additions are self-documenting, i.e. that another
  developer (someone like the reviewer) will be able to understand what the code
  does in the future. (only if we decide to keep poetry)


## PR checklist

- ~[ ] Add or expand tests; coverage checks both ✅~ Not adding functionality.
- ~[ ] Add, expand, or update documentation.~ Only if we keep poetry.
- ~[ ] Update doc/whatsnew.~ Only if we keep poetry.
